Forge.TreeWalker/src/TreeWalkerSession.cs

```diff
@@ -459,22 +498,73 @@ internal async Task PerformLeafTypeBehavior(TreeNode treeNode)
             }
         }
 
+        /// <summary>
+        /// Performs Subroutine TreeNodeType behavior.
+        /// </summary>
+        /// <param name="treeNode">The TreeNode containing actions to execute.</param>
+        /// <param name="treeNodeKey">The TreeNode's key where the actions are taking place.</param>
+        /// <exception cref="ArgumentException">If the pre-checks fail.</exception>
+        internal async Task PerformSubroutineTypeBehavior(TreeNode treeNode, string treeNodeKey)
```

Comment:
>PerformSubroutineTypeBehavior [](start = 28, length = 29)

looks like this check function can be part of PerformActionTypeBehavior()

bool preCheck_ContainsAtLeastOneSubroutineAction = false;
bool preCheck_ContainsNoSubroutineActions = true;
...

if (treeNode.Type != TreeNodeType.Subroutine && !preCheck_ContainsNoSubroutineActions)
{throw...}

if (treeNode.Type == TreeNodeType.Subroutine && !ContainsAtLeastOneSubroutineAction )
{throw...}

Reply:
I want to give the TreeNodeTypes their own methods for extra clarity. If the logic was intertwined inside ActionTypeBehavior, it becomes more difficult to know what is special about Subroutine type.
This also allows to more easily add additional logic later that is specific for certain types.

---
In reply to: [356350218](https://github.com/microsoft/Forge/pull/6#discussion_r356350218) [](ancestors = 356350218)